rfcs/20220610-saved-model-fingerprinting.md

```diff
@@ -0,0 +1,293 @@
+   </td>
+   <td>Same
+   </td>
+   <td>Same
+   </td>
+  </tr>
+</table>
+
+
+## Design Proposal
+
+The engineering impact of this project is significant in that it adds an additional file (`fingerprint.pb`) to the SavedModel format.
+
+
+### A Separate File for the Fingerprint
+
+We plan to add an additional protobuf in the SavedModel directory:
+
+*   `assets/`
+*   `saved_model.pb`
+*   **<code>fingerprint.pb</code>**
+*   <code>variables/</code>
+
+The advantages of having a separate file for the fingerprint are:
+- It neatly encapsulates the changes that this design proposes
+- It does not interfere with the MetaGraphDef in the saved_model.pb. The MetaGraphDef protobuf is already extremely large so we want to avoid adding fields to it.
+
+One drawback of having a separate file is that it is a potentially significant change to the SavedModel format and could break any users who are using the SavedModel format in ways not endorsed by our library.
+
+### API for reading and writing the fingerprints
+
+All fingerprints will be written to disk automatically, and there is no way to bypass or customize writing fingerprints when serializing a SavedModel.
+
+To read the fingerprint, users can load the protobuf into memory and read the field. At first, there will not be dedicated APIs in Python to get the fingerprint, since no other part of the SavedModel has special APIs to access its value. However, if the need arises, we can add APIs under the `tf.saved_model` namespace to read the fingerprint.
+
+### Validation
+
+When loading the SavedModel, we will log a warning (absl logging) if the fingerprint does not match the contents of `saved_model.pb`.
+
+It should not be possible to create an invalid fingerprint if clients use our APIs correctly. The fingerprint and the SavedModel
+should always be in sync. If there’s a mismatch, then we assume that someone has changed the `fingerprint.pb` or `saved_model.pb` in ways not endorsed by TensorFlow.
+
+The SavedModel CLI is also a good place for verification as well. We can check that the fingerprint is valid for the corresponding `saved_model.pb` file, based on an option in the CLI. We can display (in colored text) the result of the match.
+
+We can also explore the option of adding a Python API such as `tf.saved_model.check_fingerprint(file_name)` to enable clients to validate the SavedModel.
+
+### Implementation Details
+
+The fingerprinting will be added as a final step of serialization. The fingerprint infrastructure will take as input the MetaGraphDef and hash each component of it, doing the necessary regularization as well.
+
+The hashing will occur in C++, since it is easier to find an open source hashing algorithm in C++ than in Python. Hashing in C++ is also much faster. There will be pybinded wrappers for access from Python. This will require passing the protobuf across the Python and C++ boundary which may require an extra cycle of de/serialization.
```

Comment:
https://docs.python.org/3/library/hashlib.html 

it mainly rely on cpython and native c++ libraries.

Reply:
Ok but I think that we could remove the claim that It Is much faster in c++ if we have not profiled It.

Reply:
Done, removed that sentence.